TrecQA/overlap_features.py

```diff
@@ -0,0 +1,172 @@
+      labels.append(int(line.strip())) 
+  return qids, questions, answers, labels
+
+def compute_overlap_features(questions, answers, word2df=None, stoplist=None):
+  word2df = word2df if word2df else {}
+  stoplist = stoplist if stoplist else set()
+  feats_overlap = []
+  for question, answer in zip(questions, answers):
+    # q_set = set(question)
+    # a_set = set(answer)
+    q_set = set([q for q in question if q not in stoplist])
+    a_set = set([a for a in answer if a not in stoplist])
+    word_overlap = q_set.intersection(a_set)
+    # overlap = float(len(word_overlap)) / (len(q_set) * len(a_set) + 1e-8)
+    if len(q_set) == 0 and len(a_set) == 0:
+      overlap = 0
+    else:
+      overlap = float(len(word_overlap)) / (len(q_set) + len(a_set))
+
+
+    word_overlap = q_set.intersection(a_set)
+    df_overlap = 0.0
+    for w in word_overlap:
+      df_overlap += word2df[w]
+    if len(q_set) == 0 and len(a_set) == 0:
+      df_overlap = 0
+    else:
+      df_overlap /= (len(q_set) + len(a_set))
+
+    feats_overlap.append(np.array([
+                         overlap,
+                         df_overlap,
+                         ]))
+  return np.array(feats_overlap)
+
+def compute_overlap_idx(questions, answers, stoplist, q_max_sent_length, a_max_sent_length):
+  stoplist = stoplist if stoplist else []
+  feats_overlap = []
+  q_indices, a_indices = [], []
+  for question, answer in zip(questions, answers):
+    q_set = set([q for q in question if q not in stoplist])
+    a_set = set([a for a in answer if a not in stoplist])
+    word_overlap = q_set.intersection(a_set)
+
+    q_idx = np.ones(q_max_sent_length) * 2
+    for i, q in enumerate(question):
+      value = 0
+      if q in word_overlap:
+        value = 1
+      q_idx[i] = value
+    q_indices.append(q_idx)
+
+
+    a_idx = np.ones(a_max_sent_length) * 2
+    for i, a in enumerate(answer):
+      value = 0
+      if a in word_overlap:
+        value = 1
+      a_idx[i] = value
+    a_indices.append(a_idx)
+
+  q_indices = np.vstack(q_indices).astype('int32')
+  a_indices = np.vstack(a_indices).astype('int32')
+
+  return q_indices, a_indices
+
+def compute_dfs(docs):
+  word2df = defaultdict(float)
+  for doc in docs:
+    for w in set(doc):
+      word2df[w] += 1.0
+  num_docs = len(docs)
+  for w, value in word2df.items():
+    word2df[w] /= np.math.log(num_docs / value)
+
+  return word2df
+
+if __name__ == '__main__':
+  # ap = argparse.ArgumentParser(description="compute overlap features for SM model")
+  # ap.add_argument("dataset", help="path/to/dataset-directory", default="./TrecQA/")
+  # args = ap.parse_args()
+
+  stoplist = set([line.strip() for line in open('stopwords.txt')])
+  
+  punct = set(string.punctuation)
+  stoplist.update(punct) 
+  #stoplist = None 
+ 
+  all_questions, all_answers, all_qids = [], [], []
+  base_dir = './'
+  # base_dir = '../../data/' + sys.argv[1] + '/'
+
+  sub_dirs = ['train/','train-all/', 'raw-dev/', 'raw-test/']
+
```

Comment:
The Clean version (Wang and Ittycheriah et al. 2015, Tan et al. 2015, dos Santos et al. 2016, Wang et al. 2016) removed questions with no answers or with only positive/negative answers, thus has only 65 questions in the development set and 68 questions in the test set. For ACL QA wiki page.